Refactor API views

```
./boogiestats/boogie_api/views.py:84:1: C901 'player_scores' is too complex (10)
./boogiestats/boogie_api/views.py:145:1: C901 'player_leaderboards' is too complex (11)
./boogiestats/boogie_api/views.py:212:1: C901 'score_submit' is too complex (21)
```

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -55,8 +55,6 @@ jobs:
       - name: Run flake8
         run: |
           flake8
-        # flake fails at the moment because of the over-the-roof cyclomatic complexity in views
-        continue-on-error: true
 
   djlint:
     runs-on: ubuntu-latest

--- a/boogiestats/boogie_api/models.py
+++ b/boogiestats/boogie_api/models.py
@@ -1,6 +1,7 @@
 import json
 from hashlib import sha256
 from pathlib import Path
+from typing import Optional
 
 from django.conf import settings
 from django.contrib.auth.models import User
@@ -128,6 +129,11 @@ class Song(models.Model):
         except Song.DoesNotExist:
             raise Managed404Error("Requested song does not exist.")
 
+    def set_ranked(self, is_ranked):
+        if is_ranked and not self.gs_ranked:
+            self.gs_ranked = True
+            self.save()
+
 
 class Player(models.Model):
     objects = PlayerManager()
@@ -148,7 +154,7 @@ class Player(models.Model):
         return super().save(*args, **kwargs)
 
     @staticmethod
-    def get_by_gs_api_key(gs_api_key):
+    def get_by_gs_api_key(gs_api_key) -> Optional["Player"]:
         api_key = Player.gs_api_key_to_bs_api_key(gs_api_key)
         return Player.objects.filter(api_key=api_key).first()
 

--- a/boogiestats/boogie_api/test/test_api.py
+++ b/boogiestats/boogie_api/test/test_api.py
@@ -41,7 +41,7 @@ def test_player_scores_given_groovestats_unranked_song_that_we_dont_track(
     }
     requests_mock.get(GROOVESTATS_ENDPOINT + "/player-scores.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": gs_api_key,
     }
     response = client.get(
         "/player-scores.php",
@@ -160,7 +160,7 @@ def test_player_leaderboards_given_groovestats_unranked_song_that_we_dont_track(
         text=json.dumps(unranked_song),
     )
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": gs_api_key,
     }
     response = client.get(
         "/player-leaderboards.php",
@@ -214,7 +214,7 @@ def test_player_leaderboards_given_groovestats_ranked_song(client, gs_api_key, r
     }
     requests_mock.get(GROOVESTATS_ENDPOINT + "/player-leaderboards.php", text=json.dumps(ranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": gs_api_key,
     }
     response = client.get(
         "/player-leaderboards.php",
@@ -270,7 +270,7 @@ def test_score_submit_given_groovestats_ranked_song(client, gs_api_key, requests
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(expected_result))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}={hash}&maxLeaderboardResults=3",
@@ -313,7 +313,7 @@ def test_score_submit_given_groovestats_unranked_song_that_we_dont_track_yet(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=76957dd1f96f764e&maxLeaderboardResults=3",
@@ -372,7 +372,7 @@ def test_score_submit_given_groovestats_unranked_song_and_better_score(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -437,7 +437,7 @@ def test_score_submit_given_groovestats_unranked_song_and_worse_score(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -502,7 +502,7 @@ def test_score_submit_with_empty_comment(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -537,7 +537,7 @@ def test_score_submit_without_a_comment(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -576,7 +576,7 @@ def test_event_score_submit(
 
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(gs_response))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=afc954d593fd8bbd&maxLeaderboardResults=3",
@@ -615,7 +615,7 @@ def test_event_player_scores(
 
     requests_mock.get(GROOVESTATS_ENDPOINT + "/player-scores.php", text=json.dumps(gs_response))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.get(
         f"/player-scores.php?chartHashP{player_index}=afc954d593fd8bbd&maxLeaderboardResults=13",
@@ -647,7 +647,7 @@ def test_event_player_leaderboards(
 
     requests_mock.get(GROOVESTATS_ENDPOINT + "/player-leaderboards.php", text=json.dumps(gs_response))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.get(
         f"/player-leaderboards.php?chartHashP{player_index}=afc954d593fd8bbd&maxLeaderboardResults=13",
@@ -674,7 +674,7 @@ def test_score_submit_with_some_judgments(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -741,7 +741,7 @@ def test_score_submit_with_all_judgments(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -812,7 +812,7 @@ def test_score_submit_with_cmod_info(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -849,7 +849,7 @@ def test_score_submit_with_no_cmod_info_but_cmod_comment(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -885,7 +885,7 @@ def test_score_submit_with_no_cmod_info(
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": some_player_gs_api_key,
     }
     response = client.post(
         f"/score-submit.php?chartHashP{player_index}=0123456789ABCDEF&maxLeaderboardResults=3",
@@ -924,8 +924,8 @@ def test_score_submit_with_two_players_playing_the_same_unranked_song(client, gs
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_songs))
     kwargs = {
-        "HTTP_X_Api_Key_Player_1": "abcdef0123456789" * 4,
-        "HTTP_X_Api_Key_Player_2": "abcdef0123456789"[::-1] * 4,
+        "HTTP_x_api_key_player_1": "abcdef0123456789" * 4,
+        "HTTP_x_api_key_player_2": "abcdef0123456789"[::-1] * 4,
     }
     response = client.post(
         "/score-submit.php?chartHashP1=aaaaadd1f96f764e&chartHashP2=aaaaadd1f96f764e&maxLeaderboardResults=3",
@@ -1032,8 +1032,8 @@ def test_score_submit_with_two_players_playing_the_same_ranked_song(client, gs_a
     }
     requests_mock.post(GROOVESTATS_ENDPOINT + "/score-submit.php", text=json.dumps(unranked_songs))
     kwargs = {
-        "HTTP_X_Api_Key_Player_1": "abcdef0123456789" * 4,
-        "HTTP_X_Api_Key_Player_2": "abcdef0123456789"[::-1] * 4,
+        "HTTP_x_api_key_player_1": "abcdef0123456789" * 4,
+        "HTTP_x_api_key_player_2": "abcdef0123456789"[::-1] * 4,
     }
     response = client.post(
         "/score-submit.php?chartHashP1=aaaaadd1f96f764e&chartHashP2=aaaaadd1f96f764e&maxLeaderboardResults=3",

--- a/boogiestats/boogie_api/test/test_api.py
+++ b/boogiestats/boogie_api/test/test_api.py
@@ -1,11 +1,12 @@
 import json
+from unittest.mock import Mock
 
 import pytest
 import requests_mock as requests_mock_lib
 
 from boogiestats import __version__ as boogiestats_version
 from boogiestats.boogie_api.models import Song, Player, Score
-from boogiestats.boogie_api.views import GROOVESTATS_ENDPOINT, GROOVESTATS_RESPONSES
+from boogiestats.boogie_api.views import GROOVESTATS_ENDPOINT, GROOVESTATS_RESPONSES, create_headers
 
 
 @pytest.fixture(autouse=True)
@@ -50,6 +51,7 @@ def test_player_scores_given_groovestats_unranked_song_that_we_dont_track(
 
     assert len(requests_mock.request_history) == 1
     assert requests_mock.last_request.qs[f"chartHashP{player_index}"] == [hash]
+    assert requests_mock.last_request.headers[f"x-api-key-player-{player_index}"] == gs_api_key
     assert response.json() == {
         f"player{player_index}": {
             "chartHash": hash,
@@ -91,6 +93,25 @@ def test_player_scores_given_groovestats_ranked_song(client, gs_api_key, request
     assert requests_mock.last_request.headers[f"x-api-key-player-{player_index}"] == gs_api_key
     assert requests_mock.last_request.headers["user-agent"].endswith(f"via BoogieStats/{boogiestats_version}")
     assert response.json() == ranked_song
+
+
+def test_create_headers_filters_headers():
+    request = Mock(headers={"Host": "boogie.stats", "x-api-key-player-1": "foo", "x-api-key-player-2": "bar"})
+    assert create_headers(request) == {
+        "User-Agent": "Anonymous via BoogieStats/0.0.1",
+        "x-api-key-player-1": "foo",
+        "x-api-key-player-2": "bar",
+    }
+
+
+def test_create_headers_wraps_user_agent():
+    request = Mock(headers={"User-Agent": "ITGmania/0.5.1"})
+    assert create_headers(request) == {"User-Agent": f"ITGmania/0.5.1 via BoogieStats/{boogiestats_version}"}
+
+
+def test_create_headers_adds_user_agent_when_missing():
+    request = Mock(headers={})
+    assert create_headers(request) == {"User-Agent": f"Anonymous via BoogieStats/{boogiestats_version}"}
 
 
 @pytest.fixture

--- a/boogiestats/boogie_api/test/test_api.py
+++ b/boogiestats/boogie_api/test/test_api.py
@@ -3,13 +3,14 @@ import json
 import pytest
 import requests_mock as requests_mock_lib
 
+from boogiestats import __version__ as boogiestats_version
 from boogiestats.boogie_api.models import Song, Player, Score
 from boogiestats.boogie_api.views import GROOVESTATS_ENDPOINT, GROOVESTATS_RESPONSES
 
 
 @pytest.fixture(autouse=True)
 def requests_mock():
-    with requests_mock_lib.Mocker() as mock:
+    with requests_mock_lib.mock(case_sensitive=True) as mock:
         yield mock
 
 
@@ -17,6 +18,7 @@ def test_player_scores_without_api_keys(client):
     response = client.get("/player-scores.php", data={"chartHashP1": "01234567890ABCDEF"})
 
     assert response.json() == GROOVESTATS_RESPONSES["PLAYERS_VALIDATION_ERROR"]
+    assert response.status_code == 400
 
 
 @pytest.fixture
@@ -28,9 +30,10 @@ def gs_api_key():
 def test_player_scores_given_groovestats_unranked_song_that_we_dont_track(
     client, gs_api_key, requests_mock, player_index
 ):
+    hash = "0123456789ABCDEF"
     unranked_song = {
         f"player{player_index}": {
-            "chartHash": "0123456789ABCDEF",
+            "chartHash": hash,
             "isRanked": False,
             "gsLeaderboard": [],
         }
@@ -41,13 +44,15 @@ def test_player_scores_given_groovestats_unranked_song_that_we_dont_track(
     }
     response = client.get(
         "/player-scores.php",
-        data={f"chartHashP{player_index}": "0123456789ABCDEF"},
+        data={f"chartHashP{player_index}": hash},
         **kwargs,
     )
 
+    assert len(requests_mock.request_history) == 1
+    assert requests_mock.last_request.qs[f"chartHashP{player_index}"] == [hash]
     assert response.json() == {
         f"player{player_index}": {
-            "chartHash": "0123456789ABCDEF",
+            "chartHash": hash,
             "isRanked": True,
             "gsLeaderboard": [],
         }
@@ -56,9 +61,10 @@ def test_player_scores_given_groovestats_unranked_song_that_we_dont_track(
 
 @pytest.mark.parametrize("player_index", [1, 2])
 def test_player_scores_given_groovestats_ranked_song(client, gs_api_key, requests_mock, player_index):
+    hash = "76957dd1f96f764d"
     ranked_song = {
         f"player{player_index}": {
-            "chartHash": "76957dd1f96f764d",
+            "chartHash": hash,
             "isRanked": True,
             "gsLeaderboard": [
                 {
@@ -76,10 +82,14 @@ def test_player_scores_given_groovestats_ranked_song(client, gs_api_key, request
     }
     requests_mock.get(GROOVESTATS_ENDPOINT + "/player-scores.php", text=json.dumps(ranked_song))
     kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": gs_api_key,
+        f"HTTP_x_api_key_player_{player_index}": gs_api_key,
     }
-    response = client.get("/player-scores.php", data={f"chartHashP{player_index}": "76957dd1f96f764d"}, **kwargs)
+    response = client.get("/player-scores.php", data={f"chartHashP{player_index}": hash}, **kwargs)
 
+    assert len(requests_mock.request_history) == 1
+    assert requests_mock.last_request.qs[f"chartHashP{player_index}"] == [hash]
+    assert requests_mock.last_request.headers[f"x-api-key-player-{player_index}"] == gs_api_key
+    assert requests_mock.last_request.headers["user-agent"].endswith(f"via BoogieStats/{boogiestats_version}")
     assert response.json() == ranked_song
 
 
@@ -113,21 +123,13 @@ def song(some_player, other_player):
 
 
 @pytest.mark.parametrize("player_index", [1, 2])
-def test_player_leaderboards_requires_max_leaderboard_results_param(client, gs_api_key, player_index):
-    kwargs = {
-        f"HTTP_X_Api_Key_Player_{player_index}": gs_api_key,
-    }
-    response = client.get("/player-leaderboards.php", data={f"chartHashP{player_index}": "0123456789ABCDEF"}, **kwargs)
-    assert response.json() == GROOVESTATS_RESPONSES["MISSING_LEADERBOARDS_LIMIT"]
-
-
-@pytest.mark.parametrize("player_index", [1, 2])
 def test_player_leaderboards_given_groovestats_unranked_song_that_we_dont_track(
     client, gs_api_key, requests_mock, player_index
 ):
+    hash = "0123456789ABCDEF"
     unranked_song = {
         f"player{player_index}": {
-            "chartHash": "0123456789ABCDEF",
+            "chartHash": hash,
             "isRanked": False,
             "gsLeaderboard": [],
         }
@@ -141,17 +143,22 @@ def test_player_leaderboards_given_groovestats_unranked_song_that_we_dont_track(
     }
     response = client.get(
         "/player-leaderboards.php",
-        data={f"chartHashP{player_index}": "0123456789ABCDEF", "maxLeaderboardResults": 3},
+        data={f"chartHashP{player_index}": hash, "maxLeaderboardResults": 3},
         **kwargs,
     )
 
     assert response.json() == {
         f"player{player_index}": {
-            "chartHash": "0123456789ABCDEF",
+            "chartHash": hash,
             "isRanked": True,
             "gsLeaderboard": [],
         }
     }
+    assert len(requests_mock.request_history) == 1
+    assert requests_mock.last_request.qs[f"chartHashP{player_index}"] == [hash]
+    assert requests_mock.last_request.qs["maxLeaderboardResults"] == ["3"]
+    assert requests_mock.last_request.headers[f"x-api-key-player-{player_index}"] == gs_api_key
+    assert requests_mock.last_request.headers["user-agent"].endswith(f"via BoogieStats/{boogiestats_version}")
 
 
 @pytest.mark.parametrize("player_index", [1, 2])
@@ -199,9 +206,10 @@ def test_player_leaderboards_given_groovestats_ranked_song(client, gs_api_key, r
 
 @pytest.mark.parametrize("player_index", [1, 2])
 def test_score_submit_given_groovestats_ranked_song(client, gs_api_key, requests_mock, player_index):
+    hash = "76957dd1f96f764d"
     expected_result = {
         f"player{player_index}": {
-            "chartHash": "76957dd1f96f764d",
+            "chartHash": hash,
             "isRanked": True,
             "gsLeaderboard": [
                 {
@@ -244,7 +252,7 @@ def test_score_submit_given_groovestats_ranked_song(client, gs_api_key, requests
         f"HTTP_X_Api_Key_Player_{player_index}": gs_api_key,
     }
     response = client.post(
-        f"/score-submit.php?chartHashP{player_index}=76957dd1f96f764d&maxLeaderboardResults=3",
+        f"/score-submit.php?chartHashP{player_index}={hash}&maxLeaderboardResults=3",
         data={
             f"player{player_index}": {
                 "score": 5805,
@@ -258,11 +266,16 @@ def test_score_submit_given_groovestats_ranked_song(client, gs_api_key, requests
 
     assert Song.objects.count() == 1
     song = Song.objects.first()
-    assert song.hash == "76957dd1f96f764d"
+    assert song.hash == hash
     assert song.gs_ranked is True
     assert Score.objects.count() == 1
     assert Player.objects.count() == 1
     assert response.json() == expected_result
+    assert len(requests_mock.request_history) == 1
+    assert requests_mock.last_request.qs[f"chartHashP{player_index}"] == [hash]
+    assert requests_mock.last_request.qs["maxLeaderboardResults"] == ["3"]
+    assert requests_mock.last_request.headers[f"x-api-key-player-{player_index}"] == gs_api_key
+    assert requests_mock.last_request.headers["user-agent"].endswith(f"via BoogieStats/{boogiestats_version}")
 
 
 @pytest.mark.parametrize("player_index", [1, 2])
@@ -558,6 +571,70 @@ def test_event_score_submit(
 
     assert Song.objects.count() == 1
     assert Score.objects.count() == 1
+    event_results = response.json()[f"player{player_index}"][event_key]
+    assert event_results["everything"] == "will be passed"
+
+
+@pytest.mark.parametrize("event_key", ["rpg", "itl"])
+@pytest.mark.parametrize("player_index", [1, 2])
+@pytest.mark.parametrize("is_ranked", [True, False])
+def test_event_player_scores(
+    client, some_player, other_player, requests_mock, some_player_gs_api_key, player_index, is_ranked, event_key
+):
+    gs_response = {
+        f"player{player_index}": {
+            "chartHash": "afc954d593fd8bbd",
+            event_key: {"everything": "will be passed"},
+            "gsLeaderboard": [],
+            "scoreDelta": 8041,
+            "isRanked": is_ranked,
+            "result": "improved",
+        }
+    }
+
+    requests_mock.get(GROOVESTATS_ENDPOINT + "/player-scores.php", text=json.dumps(gs_response))
+    kwargs = {
+        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+    }
+    response = client.get(
+        f"/player-scores.php?chartHashP{player_index}=afc954d593fd8bbd&maxLeaderboardResults=13",
+        **kwargs,
+    )
+
+    assert Song.objects.count() == 0
+    assert Score.objects.count() == 0
+    event_results = response.json()[f"player{player_index}"][event_key]
+    assert event_results["everything"] == "will be passed"
+
+
+@pytest.mark.parametrize("event_key", ["rpg", "itl"])
+@pytest.mark.parametrize("player_index", [1, 2])
+@pytest.mark.parametrize("is_ranked", [True, False])
+def test_event_player_leaderboards(
+    client, some_player, other_player, requests_mock, some_player_gs_api_key, player_index, is_ranked, event_key
+):
+    gs_response = {
+        f"player{player_index}": {
+            "chartHash": "afc954d593fd8bbd",
+            event_key: {"everything": "will be passed"},
+            "gsLeaderboard": [],
+            "scoreDelta": 8041,
+            "isRanked": is_ranked,
+            "result": "improved",
+        }
+    }
+
+    requests_mock.get(GROOVESTATS_ENDPOINT + "/player-leaderboards.php", text=json.dumps(gs_response))
+    kwargs = {
+        f"HTTP_X_Api_Key_Player_{player_index}": some_player_gs_api_key,
+    }
+    response = client.get(
+        f"/player-leaderboards.php?chartHashP{player_index}=afc954d593fd8bbd&maxLeaderboardResults=13",
+        **kwargs,
+    )
+
+    assert Song.objects.count() == 0
+    assert Score.objects.count() == 0
     event_results = response.json()[f"player{player_index}"][event_key]
     assert event_results["everything"] == "will be passed"
 

--- a/boogiestats/boogie_api/views.py
+++ b/boogiestats/boogie_api/views.py
@@ -167,7 +167,7 @@ def player_scores(request):
 
         gs_player = gs_response.get(player_id, {})
         gs_leaderboard = gs_player.get("gsLeaderboard", [])
-        leaderboard = gs_leaderboard or get_local_leaderboard(player, num_entries=1)  # TODO 1? not from req?
+        leaderboard = gs_leaderboard or get_local_leaderboard(player, num_entries=1)  # limit is not present in request
 
         final_response[player_id]["gsLeaderboard"] = leaderboard
 

--- a/boogiestats/boogie_api/views.py
+++ b/boogiestats/boogie_api/views.py
@@ -77,8 +77,15 @@ def parse_players(request):
 
 
 def create_headers(request):
+    """
+    Creates headers to be used when passing requests to GS.
+
+    Some headers have to be filtered out (e.g. `Host`), otherwise GS front rejects the request.
+    It's enough to just keep API keys. We also add BoogieStats to User-Agent in case GS ever decides to make usage stats.
+    """
+
     return {
-        **request.headers,
+        **{k: v for k, v in request.headers.items() if k.lower().startswith("x-api-key")},
         "User-Agent": f"{request.headers.get('User-Agent', 'Anonymous')} via BoogieStats/{boogiestats_version}",
     }
 

--- a/boogiestats/boogie_api/views.py
+++ b/boogiestats/boogie_api/views.py
@@ -4,7 +4,7 @@ import uuid
 from collections import defaultdict
 
 import requests
-from django.http import JsonResponse, HttpResponseServerError
+from django.http import JsonResponse
 from django.views.decorators.csrf import csrf_exempt
 
 from boogiestats import __version__ as boogiestats_version
@@ -42,6 +42,7 @@ GROOVESTATS_RESPONSES = {
     },
 }
 GROOVESTATS_TIMEOUT = 12
+SUPPORTED_EVENTS = ("rpg", "itl")
 
 
 def new_session(request):  # TODO add proxying to groovestats / caching?
@@ -77,41 +78,72 @@ def parse_players(request):
 
 def create_headers(request):
     return {
+        **request.headers,
         "User-Agent": f"{request.headers.get('User-Agent', 'Anonymous')} via BoogieStats/{boogiestats_version}",
     }
 
 
+def fill_event_leaderboards(final_response, gs_player, player_id):
+    for event in SUPPORTED_EVENTS:
+        if event_leaderboard := gs_player.get(event):
+            final_response[player_id][event] = event_leaderboard
+
+
+def handle_score_results(player, gs_player, old_score, new_score_value):
+    is_ranked = gs_player.get("isRanked", False)
+    if is_ranked:
+        player["result"] = gs_player["result"]
+        player["delta"] = gs_player["scoreDelta"]
+    else:
+        if old_score:
+            if old_score.score < new_score_value:
+                player["result"] = "improved"
+            else:
+                player["result"] = "score-not-improved"
+
+            player["delta"] = new_score_value - old_score.score
+        else:
+            player["result"] = "score-added"
+            player["delta"] = new_score_value
+
+
+def get_local_leaderboard(player, num_entries):
+    gs_api_key = player["gsApiKey"]
+    chart_hash = player["chartHash"]
+    player_instance = Player.get_by_gs_api_key(gs_api_key)
+    song = Song.objects.filter(hash=chart_hash).first()
+
+    if song:
+        return song.get_leaderboard(num_entries=num_entries, player=player_instance)
+    return []
+
+
+def get_or_create_player(gs_api_key):
+    player_instance = Player.get_by_gs_api_key(gs_api_key)
+    if not player_instance:
+        machine_tag = uuid.uuid4().hex[:4].upper()
+        player_instance = Player.objects.create(gs_api_key=gs_api_key, machine_tag=machine_tag)
+    return player_instance
+
+
 def player_scores(request):
-    params = {}
     headers = create_headers(request)
 
     try:
         players = parse_players(request)
     except ValueError:
-        return JsonResponse(data=GROOVESTATS_RESPONSES["PLAYERS_VALIDATION_ERROR"])
-
-    for player_index, player in players.items():
-        gs_api_key = player["gsApiKey"]
-        chart_hash = player["chartHash"]
-        player_instance = Player.get_by_gs_api_key(gs_api_key)
-        song = Song.objects.filter(hash=chart_hash).first()
-
-        if song:
-            player["leaderboard"] = song.get_leaderboard(num_entries=1, player=player_instance)
-
-        params[f"chartHashP{player_index}"] = chart_hash
-        headers[f"x-api-key-player-{player_index}"] = gs_api_key
+        return JsonResponse(data=GROOVESTATS_RESPONSES["PLAYERS_VALIDATION_ERROR"], status=400)
 
     try:
         gs_response = requests.get(
             GROOVESTATS_ENDPOINT + "/player-scores.php",
-            params=params,
+            params=request.GET,
             headers=headers,
             timeout=GROOVESTATS_TIMEOUT,
         ).json()
     except requests.Timeout as e:
         logger.error(f"Request to GrooveStats timed-out: {e}")
-        return HttpResponseServerError()
+        return JsonResponse(GROOVESTATS_RESPONSES["GROOVESTATS_DEAD"], status=504)
 
     logger.info(gs_response)
 
@@ -126,59 +158,37 @@ def player_scores(request):
             "isRanked": True,
         }
 
-        gs_leaderboard = gs_response.get(f"player{player_index}", {}).get("gsLeaderboard", [])
-        leaderboard = gs_leaderboard or player.get("leaderboard", [])
+        gs_player = gs_response.get(player_id, {})
+        gs_leaderboard = gs_player.get("gsLeaderboard", [])
+        leaderboard = gs_leaderboard or get_local_leaderboard(player, num_entries=1)  # TODO 1? not from req?
 
         final_response[player_id]["gsLeaderboard"] = leaderboard
 
-        player["itl"] = gs_response.get(player_id, {}).get("itl")
-        if player["itl"]:
-            final_response[player_id]["itl"] = player["itl"]
-
-        player["rpg"] = gs_response.get(player_id, {}).get("rpg")
-        if player["rpg"]:
-            final_response[player_id]["rpg"] = player["rpg"]
+        fill_event_leaderboards(final_response, gs_player, player_id)
 
     return JsonResponse(data=final_response)
 
 
 def player_leaderboards(request):
-    params = {}
     headers = create_headers(request)
 
     try:
         players = parse_players(request)
     except ValueError:
-        return JsonResponse(data=GROOVESTATS_RESPONSES["PLAYERS_VALIDATION_ERROR"])
+        return JsonResponse(data=GROOVESTATS_RESPONSES["PLAYERS_VALIDATION_ERROR"], status=400)
 
-    max_results = int(request.GET.get("maxLeaderboardResults", 0))
-
-    if not max_results:
-        return JsonResponse(data=GROOVESTATS_RESPONSES["MISSING_LEADERBOARDS_LIMIT"])
-    params["maxLeaderboardResults"] = max_results
-
-    for player_index, player in players.items():
-        gs_api_key = player["gsApiKey"]
-        chart_hash = player["chartHash"]
-        player_instance = Player.get_by_gs_api_key(gs_api_key)
-        song = Song.objects.filter(hash=chart_hash).first()
-
-        if song:
-            player["leaderboard"] = song.get_leaderboard(num_entries=max_results, player=player_instance)
-
-        params[f"chartHashP{player_index}"] = chart_hash
-        headers[f"x-api-key-player-{player_index}"] = gs_api_key
+    max_results = int(request.GET.get("maxLeaderboardResults", 1))
 
     try:
         gs_response = requests.get(
             GROOVESTATS_ENDPOINT + "/player-leaderboards.php",
-            params=params,
+            params=request.GET,
             headers=headers,
             timeout=GROOVESTATS_TIMEOUT,
         ).json()
     except requests.Timeout as e:
         logger.error(f"Request to GrooveStats timed-out: {e}")
-        return HttpResponseServerError()
+        return JsonResponse(GROOVESTATS_RESPONSES["GROOVESTATS_DEAD"], status=504)
     logger.info(gs_response)
 
     final_response = {}
@@ -192,150 +202,92 @@ def player_leaderboards(request):
             "isRanked": True,
         }
 
-        gs_leaderboard = gs_response.get(f"player{player_index}", {}).get("gsLeaderboard", [])
-        leaderboard = gs_leaderboard or player.get("leaderboard", [])
+        gs_player = gs_response.get(player_id, {})
+        gs_leaderboard = gs_player.get("gsLeaderboard", [])
+        leaderboard = gs_leaderboard or get_local_leaderboard(player, max_results)
 
         final_response[player_id]["gsLeaderboard"] = leaderboard
 
-        player["itl"] = gs_response.get(player_id, {}).get("itl")
-        if player["itl"]:
-            final_response[player_id]["itl"] = player["itl"]
-
-        player["rpg"] = gs_response.get(player_id, {}).get("rpg")
-        if player["rpg"]:
-            final_response[player_id]["rpg"] = player["rpg"]
+        fill_event_leaderboards(final_response, gs_player, player_id)
 
     return JsonResponse(data=final_response)
 
 
 @csrf_exempt
 def score_submit(request):
-    params = {}
     headers = create_headers(request)
 
     try:
         players = parse_players(request)
-    except ValueError:
-        return JsonResponse(data=GROOVESTATS_RESPONSES["PLAYERS_VALIDATION_ERROR"])
+        body_parsed = json.loads(request.body)
+    except (ValueError, json.JSONDecodeError):
+        return JsonResponse(data=GROOVESTATS_RESPONSES["PLAYERS_VALIDATION_ERROR"], status=400)
 
-    max_results = int(request.GET.get("maxLeaderboardResults", 0))
-
-    if not max_results:
-        return JsonResponse(data=GROOVESTATS_RESPONSES["MISSING_LEADERBOARDS_LIMIT"])
-    params["maxLeaderboardResults"] = max_results
+    max_results = int(request.GET.get("maxLeaderboardResults", 1))
 
     try:
-        body_parsed = json.loads(request.body)
-    except json.JSONDecodeError:
-        # TODO early exit?
-        body_parsed = {}
+        gs_response = requests.post(
+            GROOVESTATS_ENDPOINT + "/score-submit.php",
+            params=request.GET,
+            headers=headers,
+            json=body_parsed,
+            timeout=GROOVESTATS_TIMEOUT,
+        ).json()
+    except requests.Timeout as e:
+        logger.error(f"Request to GrooveStats timed-out: {e}")
+        return JsonResponse(GROOVESTATS_RESPONSES["GROOVESTATS_DEAD"], status=504)
+    logger.info(gs_response)
 
-    for player_index, player in players.items():
-        gs_api_key = player["gsApiKey"]
-        chart_hash = player["chartHash"]
-
-        params[f"chartHashP{player_index}"] = chart_hash
-        headers[f"x-api-key-player-{player_index}"] = gs_api_key
-
-    if params and headers and body_parsed:
-        try:
-            gs_response = requests.post(
-                GROOVESTATS_ENDPOINT + "/score-submit.php",
-                params=params,
-                headers=headers,
-                json=body_parsed,
-                timeout=GROOVESTATS_TIMEOUT,
-            ).json()
-        except requests.Timeout as e:
-            logger.error(f"Request to GrooveStats timed-out: {e}")
-            return HttpResponseServerError()
-        logger.info(gs_response)
-    else:
-        return JsonResponse(GROOVESTATS_RESPONSES["GROOVESTATS_DEAD"])
-
-    for player_index, player in players.items():
-        player_id = f"player{player_index}"
-        chart_hash = player["chartHash"]
-        gs_api_key = player["gsApiKey"]
-        is_ranked = gs_response.get(player_id, {}).get("isRanked", False)
-
-        song, song_created = Song.objects.get_or_create(hash=chart_hash)
-        player["song"] = song
-        player["is_ranked"] = is_ranked
-
-        player_instance = Player.get_by_gs_api_key(gs_api_key)
-
-        if not player_instance:
-            machine_tag = uuid.uuid4().hex[:4].upper()
-            player_instance = Player.objects.create(gs_api_key=gs_api_key, machine_tag=machine_tag)
-
-        player["player_instance"] = player_instance
-
-        _, old_score = song.get_highscore(player_instance)
-
-        comment = body_parsed[player_id].get("comment", "")
-        used_cmod = body_parsed[player_id].get("usedCmod", None)
-        judgments = body_parsed[player_id].get("judgmentCounts", None)
-
-        player_instance.scores.create(
-            song=song,
-            score=body_parsed[player_id]["score"],
-            comment=comment,
-            rate=body_parsed[player_id].get("rate", 100),
-            used_cmod=used_cmod,
-            judgments=judgments,
-        )
-
-        player_response = gs_response.get(player_id, {})
-
-        # event structures can be present even when song is not ranked, so we just pass them when present
-        if "itl" in player_response:
-            player["itl"] = player_response["itl"]
-
-        if "rpg" in player_response:
-            player["rpg"] = player_response["rpg"]
-
-        if is_ranked:
-            player["leaderboard"] = player_response["gsLeaderboard"]
-            player["result"] = player_response["result"]
-            player["delta"] = player_response["scoreDelta"]
-
-            if not song.gs_ranked:
-                song.gs_ranked = True
-                song.save()
-        else:
-            if old_score:
-                if old_score.score < body_parsed[player_id]["score"]:
-                    player["result"] = "improved"
-                else:
-                    player["result"] = "score-not-improved"
-
-                player["delta"] = body_parsed[player_id]["score"] - old_score.score
-            else:
-                player["result"] = "score-added"
-                player["delta"] = body_parsed[player_id]["score"]
-
-    for player in players.values():
-        if not player["is_ranked"]:
-            player["leaderboard"] = player["song"].get_leaderboard(
-                num_entries=max_results, player=player["player_instance"]
-            )
+    handle_scores(body_parsed, gs_response, players)
 
     final_response = {}
 
     for player_index, player in players.items():
-        final_response[f"player{player_index}"] = {
+        player_id = f"player{player_index}"
+        gs_player = gs_response.get(player_id, {})
+        gs_leaderboard = gs_player.get("gsLeaderboard", [])
+        leaderboard = gs_leaderboard or get_local_leaderboard(player, max_results)
+
+        final_response[player_id] = {
             "chartHash": player["chartHash"],
             "isRanked": True,
-            "gsLeaderboard": player["leaderboard"],
+            "gsLeaderboard": leaderboard,
             "scoreDelta": player["delta"],
             "result": player["result"],
         }
-
-        if player.get("itl"):
-            final_response[f"player{player_index}"]["itl"] = player["itl"]
-
-        if player.get("rpg"):
-            final_response[f"player{player_index}"]["rpg"] = player["rpg"]
+        fill_event_leaderboards(final_response, gs_player, player_id)
 
     return JsonResponse(data=final_response)
+
+
+def handle_scores(body_parsed, gs_response, players):
+    for player_index, player in players.items():
+        player_id = f"player{player_index}"
+        chart_hash = player["chartHash"]
+        gs_api_key = player["gsApiKey"]
+        gs_player = gs_response.get(player_id, {})
+        is_ranked = gs_player.get("isRanked", False)
+
+        song, _ = Song.objects.get_or_create(hash=chart_hash)
+        song.set_ranked(is_ranked)
+
+        player_instance = get_or_create_player(gs_api_key)
+
+        _, old_score = song.get_highscore(player_instance)
+
+        score_submission = body_parsed[player_id]
+        comment = score_submission.get("comment", "")
+        used_cmod = score_submission.get("usedCmod", None)
+        judgments = score_submission.get("judgmentCounts", None)
+
+        player_instance.scores.create(
+            song=song,
+            score=score_submission["score"],
+            comment=comment,
+            rate=score_submission.get("rate", 100),
+            used_cmod=used_cmod,
+            judgments=judgments,
+        )
+
+        new_score_value = score_submission["score"]
+        handle_score_results(player, gs_player, old_score, new_score_value)

--- a/boogiestats/templates/boogie_ui/score_with_judgments.html
+++ b/boogiestats/templates/boogie_ui/score_with_judgments.html
@@ -3,21 +3,16 @@
     <div>{{ score.score|div:100|stringformat:".2f" }}%</div>
     {% if score.has_judgments %}
         {% with both_fantastics=score.fantastics|add:score.fantastics_plus %}
-            <div class="score-box"
-                 style="grid-template-columns: {{ both_fantastics }}fr {{ score.excellents }}fr {{ score.greats }}fr {{ score.decents }}fr {{ score.way_offs }}fr {{ score.misses }}fr;">
-                <div class="fantastic"
-                     title="{{ both_fantastics }} ({% widthratio both_fantastics score.total_steps 100 %}%)"></div>
-                <div class="excellent"
-                     title="{{ score.excellents }} ({% widthratio score.excellents score.total_steps 100 %}%)"></div>
-                <div class="great"
-                     title="{{ score.greats }} ({% widthratio score.greats score.total_steps 100 %}%)"></div>
-                <div class="decent"
-                     title="{{ score.decents }} ({% widthratio score.decents score.total_steps 100 %}%)"></div>
-                <div class="wayoff"
-                     title="{{ score.way_offs }} ({% widthratio score.way_offs score.total_steps 100 %}%)"></div>
-                <div class="miss"
-                     title="{{ score.misses }} ({% widthratio score.misses score.total_steps 100 %}%)"></div>
+            {# djlint:off #}
+            <div class="score-box" style="grid-template-columns: {{ both_fantastics }}fr {{ score.excellents }}fr {{ score.greats }}fr {{ score.decents }}fr {{ score.way_offs }}fr {{ score.misses }}fr;">
+                <div class="fantastic" title="{{ both_fantastics }} ({% widthratio both_fantastics score.total_steps 100 %}%)"></div>
+                <div class="excellent" title="{{ score.excellents }} ({% widthratio score.excellents score.total_steps 100 %}%)"></div>
+                <div class="great" title="{{ score.greats }} ({% widthratio score.greats score.total_steps 100 %}%)"></div>
+                <div class="decent" title="{{ score.decents }} ({% widthratio score.decents score.total_steps 100 %}%)"></div>
+                <div class="wayoff" title="{{ score.way_offs }} ({% widthratio score.way_offs score.total_steps 100 %}%)"></div>
+                <div class="miss" title="{{ score.misses }} ({% widthratio score.misses score.total_steps 100 %}%)"></div>
             </div>
+            {# djlint:on #}
         {% endwith %}
     {% endif %}
 </div>

